### DISCUSSION TITLE
How to set up Tiled in Angular

### DISCUSSION THREAD

**Discussion by KrzyZyb** [excaliburjs/Excalibur#2434] [Q&A] [answered]
Hi, i have a problem with paths for Tiled in my Angular ExcaliburJs project.
My project structure:

![scr1](https://user-images.githubusercontent.com/18530395/181199376-2f225b9f-2b0e-4ba8-b395-8595f6faec13.jpg)

My excalibur component (engine.component.ts) imports as follows:
`const tiledMap = new TiledMapResource("/assets/map.tmx");`

My map.tmx contains
 `path: source="map.tsx"/>`

And my map.tsx contains
` <image source="assets/tilesheet.png" width="896" height="704"/>`

In this approach i don't have any errors in browsers console but the map does not display. The hint for the reason is that if i try to edit the map directly it does not see the png file:
![scr2](https://user-images.githubusercontent.com/18530395/181200706-e93214b0-38ce-4d12-a5bb-c3ff8cf3059f.jpg)

The obvious solution would be to adjust the path to the png file in map.tsx to:
` <image source="./tilesheet.png" width="896" height="704"/>`
or just:
` <image source="tilesheet.png" width="896" height="704"/>`

When i try to do that the png resources are visible during map editing but the browser console gives me error instead:
![scr3](https://user-images.githubusercontent.com/18530395/181201958-44db702f-ed45-466c-9d47-a38a19a2a592.jpg)

UPDATE:
Changing path in map.tsx from " to image source="assets/tilesheet.png" to image source="../assets/tilesheet.png" helped to solve the resources issue, both all of them are detectable but the map itself does not load to the canvas.

**Comment by eonarheim** [comment, marked as answer]
Hi @KrzyZyb thanks for opening a discussion!

Do you have a small example repo you can share?

My first guess is the angular CLI is maybe moving files around such that the path resolution isn't working at runtime. If that's the case you can work around the Tiled plugin's path resolution by overriding the `convertPath` function similar to this https://github.com/excaliburjs/Excalibur/discussions/2370#discussioncomment-3231472

My other guess based on the first set of symptoms without modifying the tmx/tsx, is it's also possible the angular cli is mangling names, we've had problems with some bundlers mangling the constructor name or function names (this causes an internal `instanceof` check to fail and not actually add entities to the scene). https://github.com/excaliburjs/excalibur-tiled/issues/344#issuecomment-1098702780 perhaps this is happening in the angular cli?

I found this SO post that documents a way of preserving names in angular https://stackoverflow.com/a/70370992/839595

I've added an issue to look into ways of minimizing the `instanceof` impact (I know this is a problem for default Vite, but if it's also a problem with default Angular it definitely increases the priority 👍) https://github.com/excaliburjs/Excalibur/issues/2436

**Reply by eonarheim** [reply]
@KrzyZyb In your scene do you have any tilemaps?

`engine.currentScene.entities` if not, I strongly suspect the name mangling

<img width="397" alt="image" src="https://user-images.githubusercontent.com/612071/181412820-ed549355-f507-44db-8779-01d4311fb9eb.png">

**Reply by KrzyZyb** [reply]
@eonarheim thank you for your reply

This is te frontend part of the project in which angular and excaliburjs are included:
https://github.com/KrzyZyb/DnDeathmatch/tree/master/DnDeathmatchUi

answering the second part: 

> @KrzyZyb In your scene do you have any tilemaps?
> 
> `engine.currentScene.entities` if not, I strongly suspect the name mangling
> 
> <img alt="image" width="397" src="https://user-images.githubusercontent.com/612071/181412820-ed549355-f507-44db-8779-01d4311fb9eb.png">

It seems that no entities were loaded...

UPDATE:
After playing with paths, file placement and angular.json (you can see those changes im my repo) all files are properly loaded to resources and do not signalize any error in browser. Yet the engine.currentScene.entities remail empty when logging them.

**Reply by eonarheim** [reply]
@KrzyZyb good news! I was able to get your example running by updating the excalibur verisons in the package.json. I think maybe the version difference was producing the symptom I thought was caused by name mangling?

<img width="784" alt="image" src="https://user-images.githubusercontent.com/612071/181511031-21ecd54e-0af5-44d8-a299-85c7b3433c7f.png">

<img width="958" alt="image" src="https://user-images.githubusercontent.com/612071/181511134-a8f39701-d2ab-488c-8a1a-9cee3254ebaa.png">

Also if you want to set the Tiled files back to their original state so they work in Tiled, you can override the `convertPath` to change how the Excalibur Tiled plugin looks for files. `originPath` is the path to the main tmx or tsx, and `relativePath` is each referenced png file in the tmx/tsx.

```typescript
const tiledMap = new TiledMapResource("/assets/map.tmx");
// this works as long as everything is in the assets folder and no deeper
tiledMap.convertPath = (originPath, relativePath) => {
  const segments = relativePath.split('/');
  const filename = segments[segments.length - 1];
  return '/assets/' + filename;
}
```

map.tmx contains
`path: source="map.tsx"/>`

map.tsx contains
`<image source="tilesheet.png" width="896" height="704"/>`

<img width="1199" alt="image" src="https://user-images.githubusercontent.com/612071/181513427-38cfa657-decc-4b45-bd9e-fb6d86c49307.png">

**Reply by eonarheim** [reply]
@KrzyZyb I've uploaded my edits to a fork https://github.com/eonarheim/DnDeathmatch

**Reply by KrzyZyb** [reply]
My mind almost exploded when after merging your changes the entities still were empty...
Luckily it was just my amnesie because i forgot to npm install ;)

THANK YOU very much for your assistance @eonarheim, this works perfectly fine!

**Reply by eonarheim** [reply]
@KrzyZyb For sure! Glad I could help :)
